### PR TITLE
mdns: add a txt record for http & json services containing 'uuid' and 'txtvers'

### PR DESCRIPTION
**Pull request by rytilahti**
## Description
Allows distinguishing between kodi installations without resorting to making any API calls.
This is also useful as kodi sends advertisements for every bound interface, causing duplicate
advertisements from a single host.

The change follows works similarly to code below in the same file for airplay support.
'txtvers' as per RFC 6763 is added for possible future extensions (web interface path? information about required authentication?)

## Motivation and Context
Currently Kodi advertises its API endpoint on all interfaces, without a possibility to distinguish a single instance without using API calls (which are cumbersome due to responding 'Busy' in their first call, e.g., when trying to fetch the mac address.
Considering mDNS/DNS-SD has a possibility for txt records, which are already used for other advertised services (where this code is adapted), it makes sense to add this information to help upstream users (e.g. netdisco and homeassistant) to automatically discovery the instance without creating duplicates due to multiple advertisements for different IP addresses.

## How Has This Been Tested?
Tested to work with `avahi-browse -ar`:
```
= docker0 IPv4 Kodi (nuc)                                    Web Site             local
   hostname = [nuc.local]
   address = [172.17.0.1]
   port = [8080]
   txt = ["txtvers=1" "mac=F8:63:XX:XX:XX:XX"]
```
and netdisco:
```
kodi:
[{'host': '172.17.0.1',
  'hostname': 'nuc.local.',
  'mac_address': 'F8:63:XX:XX:XX:XX',
  'port': 8080,
  'properties': {'mac': 'F8:63:XX:XX:XX:XX', 'txtvers': '1'}}]
```

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by wsnipex** [comment]
how does hardcoding txtvers to 1 help distinguish different kodi instances?

**Comment by rytilahti** [comment]
So that's just a recommendation for dns-sd records: https://tools.ietf.org/html/rfc6763#section-6.7 - I thought to include it for the sake of completeness, but I can very well remove it.

**Comment by wsnipex** [comment]
ok. I have no problem with the txtvers, but as the rfc states:
>It is recommended that authors defining DNS-SD profiles include an
   attribute of the form "txtvers=x", where "x" is a decimal version
   number in US-ASCII [RFC20] text (e.g., "txtvers=1" or "txtvers=8"),
   in their definition, **and require it to be the first key/value pair in
   the TXT record.**

**Comment by Rechi** [comment]
@rytilahti please squash your second commit into the first one, we wan't a clean git history.

**Comment by davilla** [comment]
just FYI, on iOS, the MAC address is not available. Nor can you get MAC addresses of other devices at runtime. Privacy issue.

**Comment by rytilahti** [comment]
@Rechi done (btw, github has a squash & merge, which I find very nice)
@davilla good to know, however, is there some better alternative for this use case? the most devices I know simply use the mac as an unique identifier.

**Comment by davilla** [comment]
the kodi codebase already addressed that issue with a unique identifier. forget the details, there was a PR for it.

**Comment by rytilahti** [comment]
I found #10543 which never got merged. Although from the looks of it the interface listing may change its order, making the method I used here makes mac as in unique identifier unreliable (without extra steps to sort the list).

**Comment by wsnipex** [comment]
yes, using the MAC is not guaranteed to work 100%, what usually interface ordering doesn't change.
Do you still want this merged as is or look for a another solution?

**Comment by rytilahti** [comment]
Looks like I missed to add the record for the rpc interface advertisement, so I added it to the same commit. Anyway, looks like at least in my case the MAC address keeps changing (no idea why, maybe the docker initializes its interfaces differently on different boots..), so I think I need to find either another identifier or simply find a way to order the list of all available MACs and pick the first one (non-localhost) from there.

Do you have any preference how to achieve that in a nice manner?

**Comment by wsnipex** [comment]
why not use something else then a mac? We already have a setting "services.devicename". This is user configurable, but defaults to a non-unique "Kodi" string.
How about adding a uuid that is auto generated if not set?

**Comment by rytilahti** [comment]
If an uuid generated and stored to the config during the first startup would be acceptable, I suppose that would be the nicest solution for this use case. Making that happen will take some time though, as it involves more changes the patch proposed here and I'm not familiar with the code.

**Comment by wsnipex** [comment]
we discussed this a bit internally and a random uuid stored seems acceptable. Let us know if you want to tackle it.

**Comment by rytilahti** [comment]
Yes, I would like to tackle it. It would spare me quite a bit of time if you can give me some pointers where to start with it, could you help with that?
1. StringUtils.createUUID() for generation.
2. Adding a new CSettings::SETTING_SERVICES_DEVICEUUID (`services.deviceuuid`) below DEVICENAME in `settings/Settings.cpp`
3. I could not find how the defaults are generated, the UUID generation should probable happen just there?
4. Modify the patch in this PR to read the uuid (`CServiceBroker::GetSettings().GetString(CSettings::SETTING_SERVICES_DEVICEUUID);`) and use that instead of mac.
5. Does `utils/SystemInfo.cpp` and respective UI parts require changing, or is it okay to keep this internal?

Does that sound reasonable?

**Comment by wsnipex** [comment]
yes, that sounds good.
ad 2) I'm not an expert on our settings system, but something like https://github.com/xbmc/xbmc/blob/master/xbmc/settings/SettingConditions.cpp#L361 should work for calling createuuid if the setting is empty.
Maybe someone has a better idea?
ad 4) if we store in settings, it's show there and imho not needed in SystemInfo additionally

**Comment by rytilahti** [comment]
Ok! I adjusted locally everything else except the `SettingConditiion` part, is that the place or the idea where the generation should be inserted?

4) Great!

**Comment by wsnipex** [comment]
I don't have a better suggestion atm. Maybe others will chime in once there is something to review

**Comment by rytilahti** [comment]
Ok, thanks, I'll try to find a way. (Btw, there's no CApplication.cpp nor PlatformInit in the whole codebase, that sounded like a place where it probably should go though!)

**Comment by wsnipex** [comment]
There is Application.cpp that has the Class CApplication.
Anyway, looks like you could add the uuid initialization to CSettings::InitializeDefaults() in https://github.com/xbmc/xbmc/blob/master/xbmc/settings/Settings.cpp#L638

**Comment by rytilahti** [comment]
Thanks, I got a bit further (reading and writing the uuid works) but it is still incomplete. Open problems:
1. Default being empty, `SetDefault` seems to just update the existing default value so it will reset after every boot. I tried to go around this by guarding it with a condition, but considering default is empty it is not working either. Ideas?
2. The value is displayed in the UI although this is probably unwanted. Removing the `control` tag doesn't work, adding a "hidden" tag will just hide its value.

File: xbmc/settings/Settings.cpp [hunk 1]
```diff
@@ -659,6 +660,15 @@ void CSettings::InitializeDefaults()
...
   if (g_application.IsStandAlone())
     std::static_pointer_cast<CSettingInt>(GetSettingsManager()->GetSetting(CSettings::SETTING_POWERMANAGEMENT_SHUTDOWNSTATE))->SetDefault(POWERSTATE_SHUTDOWN);
+
+  // Check if a device UUID is set, and if not, set the default.
+  // Without this check this would cause re-setting it on every boot.
+  // FIXME: This does not work either. Checking for emptiness & using SetValue fails likewise. Ideas?
+  std::shared_ptr<CSettingString> deviceUUID = std::static_pointer_cast<CSettingString>(GetSettingsManager()->GetSetting(CSettings::SETTING_SERVICES_DEVICEUUID));
+  if (deviceUUID->GetValue().empty()) {
+    const std::string& uuid = StringUtils::CreateUUID();
+    std::static_pointer_cast<CSettingString>(GetSettingsManager()->GetSetting(CSettings::SETTING_SERVICES_DEVICEUUID))->SetDefault(uuid);
```
Comment (wsnipex) [inline xbmc/settings/Settings.cpp]: don't set the default here. Instead, create and set a new uuid.

File: system/settings/settings.xml [hunk 2]
```diff
@@ -1620,6 +1620,17 @@
           <control type="toggle" />
         </setting>
       </group>
+      <group id="3">
+      <requirement>HAS_ZEROCONF</requirement>
+        <setting id="services.deviceuuid" type="string">
```
Comment (Rechi) [inline system/settings/settings.xml]: add `<visible>false</visible>` inside the `setting` tag to hide it in the GUI

File: xbmc/settings/Settings.cpp (hunk 1 above)
Comment (rytilahti) [reply xbmc/settings/Settings.cpp]: Great, I'm unsure why it didn't work in the earlier try. I fixed that and force pushed the update in!

File: system/settings/settings.xml (hunk 2 above)
Comment (rytilahti) [reply system/settings/settings.xml]: Thanks rechi, that worked! I was searching previously for hidden for edit, which was of course the wrong one.

File: system/settings/settings.xml [hunk 3]
```diff
@@ -1620,6 +1620,18 @@
           <control type="toggle" />
         </setting>
       </group>
+      <group id="3">
+      <requirement>HAS_ZEROCONF</requirement>
```
Comment (wsnipex) [inline system/settings/settings.xml]: I'm unsure if we should keep this requirement or make it available globally.
@peak3d you said in slack, that you could make use of such a uuid as well?

Comment (rytilahti) [reply system/settings/settings.xml]: I can adjust it as seen necessary, it probably then would not be anymore inside `service` namespace but somewhere else if that's the case?

**Review by wsnipex — APPROVED** [review]

File: xbmc/settings/Settings.cpp [hunk 4]
```diff
@@ -659,6 +660,13 @@ void CSettings::InitializeDefaults()
 
   if (g_application.IsStandAlone())
     std::static_pointer_cast<CSettingInt>(GetSettingsManager()->GetSetting(CSettings::SETTING_POWERMANAGEMENT_SHUTDOWNSTATE))->SetDefault(POWERSTATE_SHUTDOWN);
+
+  // Initialize deviceUUID if not already set, used in zeroconf advertisements.
+  std::shared_ptr<CSettingString> deviceUUID = std::static_pointer_cast<CSettingString>(GetSettingsManager()->GetSetting(CSettings::SETTING_SERVICES_DEVICEUUID));
+  if (deviceUUID->GetValue().empty()) {
```
Comment (Rechi) [inline xbmc/settings/Settings.cpp]: curly brackets has to go into new line, see https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-braces

Comment (rytilahti) [reply xbmc/settings/Settings.cpp]: Done :+1:

**Pull request merged**